src/TokenParser/TypesTokenParser.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+/*
+ * This file is part of Twig.
+ *
+ * (c) Fabien Potencier
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Twig\TokenParser;
+
+use Twig\Error\SyntaxError;
+use Twig\ExpressionParser;
+use Twig\Node\Expression\ArrayExpression;
+use Twig\Node\Expression\ConstantExpression;
+use Twig\Node\Expression\NameExpression;
+use Twig\Node\Node;
+use Twig\Node\SetNode;
+use Twig\Node\TypesNode;
+use Twig\Token;
+use Twig\TokenStream;
+
+/**
+ * Declare variable types.
+ *
+ * The mapping must use names as keys (not strings) and type strings as values.
+ * 
+ * Names may be suffixed with `?` to mark them as optional (i.e., the context may not
+ * have that variable defined).
+ * 
+ * Example:
+ * ```
+ *    {% types {foo: 'int', bar?: 'string'} %}
+ * ```
+ * @author Jeroen Versteeg <jeroen@alisqi.com>
+ * @see    https://github.com/twigphp/Twig/issues/4165
```

Comment:
Those `@see` linking to the Twig issue requesting the feature should be removed.